From f1bf82ac8e5a488fda285e019012577c4b3d81bc Mon Sep 17 00:00:00 2001
From: Varun Sharma <vsharma@habana.ai>
Date: Wed, 30 Dec 2020 13:09:13 +0200
Subject: [PATCH] [SW-24979]: agents: fw-loader compilation support

This patch adds compilation support for fw-loader agent, and
will be built as part of "gaudi-asic-build" command.

This addition would create the following upon gaudi-asic-build:
gaudi-agent-fw_loader-fit.itb

Change-Id: I2ce45d4fdff5b078935d67d37608dc51a3ec0439
Signed-off-by: Varun Sharma <vsharma@habana.ai>
---

diff --git a/configs/gaudi_defconfig b/configs/gaudi_defconfig
index bd9d111..85a0b03 100644
--- a/configs/gaudi_defconfig
+++ b/configs/gaudi_defconfig
@@ -65,3 +65,4 @@
 BR2_PACKAGE_ZEPHYR_ARM_AGENT_HBM=y
 BR2_PACKAGE_ZEPHYR_ARM_AGENT_SERDES=y
 BR2_PACKAGE_ZEPHYR_ARM_AGENT_PCIE=y
+BR2_PACKAGE_ZEPHYR_ARM_AGENT_FW_LOADER=y
diff --git a/package/zephyr/Config.in b/package/zephyr/Config.in
index ecded4c..0dd88eb 100644
--- a/package/zephyr/Config.in
+++ b/package/zephyr/Config.in
@@ -32,3 +32,9 @@
         select BR2_PACKAGE_ZEPHYR
         help
           PCIe on Zephyr ARM
+
+config BR2_PACKAGE_ZEPHYR_ARM_AGENT_FW_LOADER
+        bool "Zephyr ARM agent fw-loader"
+        select BR2_PACKAGE_ZEPHYR
+        help
+          FW Loader using Zephyr on ARM as backend.
diff --git a/package/zephyr/zephyr.mk b/package/zephyr/zephyr.mk
index de4b6df..3108add 100644
--- a/package/zephyr/zephyr.mk
+++ b/package/zephyr/zephyr.mk
@@ -86,6 +86,18 @@
 $(call CREATE_ZEPHYR_MARGIN_TOOL_AGENT,pcie)
 endif
 
+ifdef BR2_PACKAGE_ZEPHYR_ARM_AGENT_FW_LOADER
+ZEPHYR_AGENT_DIR = build-agent-fw-loader
+ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_AGENT_DIR)
+ZEPHYR_CONF_OPTS_AGENT += -DBOARD=hl_$(DEVICE_NAME)_$(HL_TARGET)
+ZEPHYR_CONF_OPTS_AGENT += $(dir $(BR2_EXTERNAL_HABANALABS_PATH))/zephyr-plugins/fw-loader/
+ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(ZEPHYR_CONF_OPTS_AGENT) -B $(ZEPHYR_AGENT_DIR)
+ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_AGENT_DIR)
+ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_AGENT_DIR)/zephyr/zephyr.bin $(BINARIES_DIR)/zephyr_agent_fw_loader.bin
+# The stat file is used for extracting the binary's load address and start address
+ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_AGENT_DIR)/zephyr/zephyr.stat $(BINARIES_DIR)/zephyr_agent_fw_loader.stat
+endif
+
 ifeq ($(ZEPHYR_UNIT_TEST),1)
 ZEPHYR_UTEST_DIR = build-utest
 ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_UTEST_DIR)
